W-13178731-rtf-deployment-statuses-kt (#1076)
mulesoft/docs-runtime-fabric: modules/ROOT/pages/deployment-status-messages.adoc

= Deployment Status Messages for Runtime Fabric

> Either when you deploy a Mule app for the first time or you make changes to an existing Mule app deployed to Runtime Fabric in Runtime Manager, you can receive multiple deployment status messages:

Reviewer: ```suggestion
Runtime Manager shows deployment status messages when you deploy a Mule app for the first time or when you make changes to an existing Mule app deployed to Runtime Fabric:
```

* xref:runtime-manager::managing-deployed-applications.adoc#application_deploy_states[Application Deployment Status State] +
These statuses are related to the general application deployment, such as Deleted, Applying, Applied, Failed, or Unknown messages.

* xref:runtime-manager::managing-deployed-applications.adoc#replica_status_states[Replica Status States] +
These statuses appear when you apply new, or changed, configurations to a deployed Mule app.

* <<warning-status,Warning Messages Status States>> +
These statuses appear when you receive a warning message while a deployment is in progress.

* <<app-namespaces,App Namespaces Status States>> +
These statuses appear when there are issues with app namespaces during deployment.

[[warning-status]]
== Warning Messages Status States

When you deploy a Mule app or make a change to an existing Mule app deployed to Runtime Fabric in Runtime Manager, you can see warning messages. Understanding the meaning of these messages is important, as some warnings are only temporary while others can be permanent failures.

> Runtime Manager waits a minute after a Mule app is deployed or a new change is applied to an existing Mule app deployed. Runtime Manager then begins to actively check the reported status of its replicas. Any status other than STARTED, STOPPED or pulling image is considered unexpected, and a warning is displayed. This does not necessarily mean there is something wrong with your Mule app or cluster. 

Reviewer: ```suggestion
When you deploy a Mule app or make a change to an existing Mule app deployed to Runtime Fabric, Runtime Manager waits a minute and then begins to actively check the reported status of its replicas. If the status is other than STARTED or STOPPED, a warning message shows. 

A warning message does not necessarily mean there is something wrong with your Mule app or cluster. Understanding the meaning of these messages is important, as some warnings are only temporary while others can be permanent failures.

Warning messages appear with the following legend: `1 or more replicas in unexpected state:` followed by the specific description. For example:

`1 or more replicas in unexpected state: [Kubernetes] Container "any point-monitoring" - PodInitializing`

[NOTE]

> In the event the warning does not resolve by itself, work with your IT department and developer team to check the deployment and pod status for troubleshooting.

Reviewer: ```suggestion
If a warning message persists and the related issue does not resolve automatically, work with your IT department and developer team to check the deployment and pod status for troubleshooting.
```

== Permanent versus Temporary Warning Messages

> Permanent warning messages status usually requires your action to be fixed. Instead, temporary warning messages disappear by themselves. Though, it is advised that you check them with your Kubernetes admin if they don't disappear in reasonable time.

Reviewer: ```suggestion
Permanent warning messages usually require your action to fix the related issue. Instead, temporary warning messages disappear after some time. Check warning messages with your Kubernetes admin if they don't disappear in a reasonable time.
```

> For example, if a node goes down, you likely have pods pending temporarily. Your Kubernetes admin may need to bring that node up manually. Or maybe the machine is taking its time to be rebooted.

Reviewer: ```suggestion
For example, if a node stops working, some pods might be pending temporarily. Your Kubernetes admin may need to restart that node manually, or the machine might take a long time to reboot. 
```

> A Mule app deployment may return a CrashLoopBackOff warning because the Mule app fails to start due to an external dependency being down, such as a database or JMS server. The warning may fix itself when the service comes up. Though, if there is something wrong with the deployment artifact, then the warning message will be permanent.

Reviewer: ```suggestion
As another example, consider the following scenario. A Mule app deployment returns a CrashLoopBackOff warning if the Mule app fails to start due to an external dependency, such as a database or JMS server, being unavailable. The warning message disappears when the service comes up. However, if there is something wrong with the deployment artifact, then the warning message is permanent.
```

== Warning Messages Description

> The following table lists different types of warning messages. Note that this is not an exhaustive list and for the actual reference you must refer to the Kubernetes documentation.

Reviewer: ```suggestion
The following table shows a non-exhaustive list of different types of warning messages. For a complete reference, refer to the Kubernetes documentation.
```

[%header%autowidth.spread]
|===
|Warning Status|Type |Description |Example

> |PodInitializing|Usually temporary |This status means that there is a pod initialization delay.  Which means that not all containers in your Mule app pods are ready and that it is usually just a matter of time until the pods initialize. If there is a real problem with your deployment, this warning transitions into another state. |<<link-example-1,Link>>

Reviewer: ```suggestion
|PodInitializing|Usually temporary |There is a pod initialization delay, not all containers in your Mule app pods are ready and it is usually a matter of time until the pods initialize. If there is a real problem with your deployment, this warning transitions into another state. |<<link-example-1,Link>>
```

> |CrashLoopBackOff|Usually permanent |This status means there is a problem with your Mule app image and the app is crashing. |<<link-example-2,Link>>

Reviewer: ```suggestion
|CrashLoopBackOff|Usually permanent |There is a problem with your Mule app image and the app is crashing. |<<link-example-2,Link>>
```

> |Unschedulable|Usually temporary |This status means that there are no nodes available to run the requested replicas. There could be a few reasons, for example: 

Reviewer: ```suggestion
|Unschedulable|Usually temporary |There are no nodes available to run the requested replicas. Reasons might include: 
```

> * Some nodes may be down for maintenance. +
> * Nodes may not have enough resources available to assign for your deployment. +
> * The *Enforce deploying replicas across nodes* option the *Deployments Target* tab of Runtime Manager is enabled but there are not enough different nodes available for the number of replicas desired. |<<link-example-3,Link>>

Reviewer: ```suggestion
* Nodes being down for maintenance. +
* Nodes without enough resources available to assign for your deployment. +
* The *Enforce deploying replicas across nodes* option in the *Deployments Target* tab of Runtime Manager is enabled but there are not enough nodes available for the number of replicas desired. |<<link-example-3,Link>>
```

|===

> For more information on Pod Status, refer to the https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-lifetime[Kubernetes official documentation^]. Your Kubernetes administrator is also a great resource to inquire about whether a state should be considered critical or not and advise how to troubleshoot if required.

Reviewer: ```suggestion
For more information on Pod Status, refer to the https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-lifetime[Kubernetes official documentation^]. Reach your Kubernetes administrator to inquire about whether a state should be considered critical or not and request advice on how to troubleshoot if required.
```

== Warning Messages Examples

[[link-example-1]]
=== PodInitializing Example

> In the following example, the warning message means the pod is ready but is waiting to initialize:

Reviewer: ```suggestion
In the following example, the warning message shows because the pod is ready but is waiting to initialize:
```

image::rtf-deployment-example-1.png[PodInitializing warning status in Runtime Manager UI]

[[link-example-2]]
=== CrashLoopBackOff Example

> In the following example, the warning message  means there is a problem with your Mule app image. Because a corrupt artifact was uploaded and caused the Mule app to crash during the deployment.

Reviewer: ```suggestion
In the following example, the warning message shows that there is a problem with your Mule app image because a corrupt artifact was uploaded and caused the Mule app to crash during the deployment.
```

image::rtf-deployment-example-2.png[CrashLoopBackOff warning status in Runtime Manager UI]

> Your Kubernetes administrator can run in the command line interface the following command to get the error or warning status as well: 

Reviewer: ```suggestion
Your Kubernetes administrator can run the following command in the command line interface to get the error or warning status as well: 
```

`kubectl get deployments,pods --namespace <ENVIRONMENT_NAMESPACE>`

For example: 

[source,text,linenums]
----
➜  ~ kubectl get deployments,pods --namespace 788a26d2-a837-4336-aa7b-1f367cf70d7d
NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/omg-s1-now   0/1     1            0           20m

NAME                              READY   STATUS             RESTARTS        AGE
pod/omg-s1-now-78b7ffd454-h5x5n   1/2     CrashLoopBackOff   8 (2m57s ago)   20m
----

[[link-example-3]]
=== Unschedulable Example

> In the following example, the warning message means that 3 nodes were unschedulable because they were not available:

Reviewer: ```suggestion
In the following example, the warning message shows that 3 nodes were unschedulable because they were not available:
```

image::rtf-deployment-example-2.png[Unschedulable warning status in Runtime Manager UI]

> Your Kubernetes administrator can run in the command line interface the following command to get the error or warning status as well: 

Reviewer: ```suggestion
Your Kubernetes administrator can run the following command in the command line interface to get the error or warning status as well: 
```

`kubectl get deployments,pods --namespace <ENVIRONMENT_NAMESPACE>`

For example: 

[source,text,linenums]
----
NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/omg-s1-now   0/1     1            0           170m

NAME                              READY   STATUS    RESTARTS   AGE
pod/omg-s1-now-58dd54d9f-xt569    0/2     Pending   0          4s
pod/omg-s1-now-6c5c5d987c-fbmbl   0/2     Pending   0          50m
----

== App Namespaces Status States

> The following table lists statuses when there are issues with app and authorized namespaces. They are prefixed with`[Runtime Fabric]`.

Reviewer: ```suggestion
The following table lists statuses that show when there are issues with the app and authorized namespaces. They are prefixed with`[Runtime Fabric]`.
```

[%header%autowidth.spread]
|===
|Status and description | Type

> |There is an existing namespace <app_namespace> for your current environment <env_id>, please create a custom namespace with required labels to proceed with this deployment. For  more information on creating custom namespaces, refer to documentation. | Permanent

Reviewer: ```suggestion
|There is an existing namespace <app_namespace> for your current environment <env_id>, please create a custom namespace with required labels to proceed with this deployment. For more information on creating custom namespaces, refer to documentation. | Permanent
```
